lib/redirect-pool.js

```diff
@@ -0,0 +1,153 @@
+const { Readable } = require('stream')
+const { pipeline, request, stream } = require('./agent')
+const util = require('./core/util')
+const Pool = require('./pool')
+
+const kOrigin = Symbol('Origin')
+const kDispatchMethod = Symbol('RedirectPoolDispatchMethod')
+const kRedirectsLeft = Symbol('RedirectPoolRedirectsLeft')
+
+const redirectCodes = [300, 301, 302, 303, 307, 308]
+const defaultMaxRedirections = 10
+
+class RedirectDiscardedResponse extends Readable {
+  constructor(resume) {
+    super({ autoDestroy: true, read: resume })
+  }
+}
+
+class RedirectPool extends Pool {
+  constructor(origin, options) {
+    super(origin, options)
+
+    this[kOrigin] = origin
+  }
+
+  request(opts, callback) {
+    opts[kDispatchMethod] = 'request'
+
+    return super.request(opts, callback)
+  }
+
+  stream(opts, factory, callback) {
+    opts[kDispatchMethod] = 'stream'
+
+    return super.stream(opts, factory, callback)
+  }
+
+  pipeline(opts, handler) {
+    opts[kDispatchMethod] = 'pipeline'
+
+    return super.pipeline(opts, handler)
+  }
+
+  dispatch(opts, handler) {
```

Comment:
Don't mutate opts.

Reply:
I have to think on this a little more.

While I can probably remove overloads of `request`, `stream` and `pipeline`, I still have to take track of of previous and redirects and left redirects.
And since when following a redirection I call the top level method, I'm completely in a stateless env.
Thus, adding few vars to opts it's my only choice, or not?
Moreover, I see that opts are already mutated when needed, let's say in `dispatchFromAgent` where agent is currently (in upstream) extracted from opts.

What do you think?

Reply:
You can just make a copy of the opts object the first time. The only important thing is that we don't mutate any object that the user passes.

Reply:
So, I've double checked the code. We're technically already doing that. 

https://github.com/nodejs/undici/blob/2366fa34e30ac5a8c33a3d02259f2ae0f7678c85/lib/agent.js#L85

Here's we are extracting `agent` and we are providing default values for `agent` and `method`.
In the current upstream `agent` is no longer passed to the pool, while in my PR it is (in order to allow recursion).

Now, since we use destructuring at the beginning of the function and rest assignment at the end, we're forwarding a new object to pool everytime, so we're not mutating the user provided `opts` object.

Moreover, when I follow a redirect, I only add two properties (via private symbols), so I'm not overwriting user provided values.
Am I wrong on this?

Reply:
This is solved now. I'm not mutating opts anymore.